Better handle parallelism in cache priming

ChayimFriedman2 opened the pull request:
Make best use of all available CPU cores.

Closes rust-lang/rust-analyzer#19711.

Also change the flags of the `prime-caches` CLI command, and make it parallel by default.

This saves 0.305ms in the parallel `prime-caches` on omicron, although the benchmarks were very noisy.

I ran into a Salsa panic while running parallel prime caches on buck2... Need to debug that, but it doesn't reproduce now.

Veykril commented on crates/ide-db/src/prime_caches.rs:
```diff
@@ -35,58 +33,112 @@ pub fn parallel_prime_caches(
+    // The idea is that if we have a def map available to compute, we should do that first.
+    // This is because def map is a dependency of both import map and symbols. So if we have
+    // e.g. a def map and a symbols, if we compute the def map we can, after it completes,
+    // compute the def maps of dependencies, the existing symbols and the symbols of the
```
```suggestion
    // compute the def maps of dependents, the existing symbols and the symbols of the
```
what is meant with existing symbols? The one we already have? Why list that given we already computed it

ChayimFriedman2 replied:
I meant the symbols we already could compute when we computed the def map.

Veykril reviewed (commented):
So the current setup doing things in phases has the problem that we only parallelize up to the width of the crate graph at the current depth which tends to be less than the number of threads available.

So the idea of this PR is to interleave the import map and crate symbols computation such that we work on those already while potentially waiting on more defmap computations to "unlock".

Is that correct?

Veykril commented on crates/ide-db/src/prime_caches.rs:
```diff
@@ -35,58 +33,112 @@ pub fn parallel_prime_caches(
+    // Such def map will just block on the dependency, which is just wasted time. So better
+    // to compute the symbols/import map of an already computed def map in that time.
+
+    let (reverse_deps, mut to_be_done_deps) = {
+        let all_crates = db.all_crates();
+        let mut reverse_deps =
+            all_crates.iter().map(|&krate| (krate, Vec::new())).collect::<FxHashMap<_, _>>();
+        let mut to_be_done_deps =
+            all_crates.iter().map(|&krate| (krate, 0u32)).collect::<FxHashMap<_, _>>();
+        for &krate in &*all_crates {
+            for dep in &krate.data(db).dependencies {
+                reverse_deps.get_mut(&dep.crate_id).unwrap().push(krate);
+                *to_be_done_deps.get_mut(&krate).unwrap() += 1;
+            }
+        }
+        (reverse_deps, to_be_done_deps)
+    };
```
```suggestion
    let (reverse_deps, mut to_be_done_deps) = {
        let all_crates = db.all_crates();
        let to_be_done_deps =
            all_crates.iter().map(|&krate| (krate, krate.data(db).dependencies.len() as u32)).collect::<FxHashMap<_, _>>();
        let mut reverse_deps =
            all_crates.iter().map(|&krate| (krate, Vec::new())).collect::<FxHashMap<_, _>>();
        for &krate in &*all_crates {
            for dep in &krate.data(db).dependencies {
                reverse_deps.get_mut(&dep.crate_id).unwrap().push(krate);
            }
        }
        (reverse_deps, to_be_done_deps)
    };
```

Veykril commented on crates/ide-db/src/prime_caches.rs:
```diff
@@ -35,58 +33,112 @@ pub fn parallel_prime_caches(
+                // Biased because we want to prefer def maps.
+                crossbeam_channel::select_biased! {
+                    recv(def_map_work_receiver) -> work => {
+                        let Ok((crate_id, crate_name)) = work else { return Ok::<_, crossbeam_channel::SendError<_>>(()) };
```
nit, could break these early returns and just `Ok::<_, crossbeam_channel::SendError<_>>(())` the end of the closure
```suggestion
                        let Ok((crate_id, crate_name)) = work else { break; };
```

Veykril commented on crates/ide-db/src/prime_caches.rs:
```diff
@@ -103,138 +155,112 @@ pub fn parallel_prime_caches(
             crates_currently_indexing: crates_currently_indexing.values().cloned().collect(),
-            crates_done,
-            crates_total,
+            crates_done: crate_def_maps_done,
+            crates_total: crate_def_maps_total,
```
Feels like we should be using the total sum numbers here and not just the def maps? Unsure, it might be a bit weird to see this be n / n in the status message without finishing yet (due to us working on symbol / import maps still).

Alternatively we could swap out the `work_type` message once we are done with the crate def maps with something else

ChayimFriedman2 replied:
Yeah, the problem with this PR is that status report gets more complicated.

Veykril reviewed (approved):
Lgtm

ChayimFriedman2 commented:
@Veykril I addressed most of your comments, but I don't know what to do with the status report given that we can't know the total amount of work as it requires knowing how module each crate has, and we can know that only after the def map completes. I agree seeing n/n crates done and it still working is weird, and if you have some idea I'd love to hear.

ChayimFriedman2 commented:
> Let's just swap the `Indexing` work type to `Collecting Symbols` once we are at `n/n`. I think that's good enough for now. We can try adjusting things once we know how this looks UX wise after merging.

Done that.
